### PR TITLE
handle pr notifications on the same thread

### PR DESCRIPTION
**Pull request by thatportugueseguy**
This PR handles PR and PR-related github notifications on monorobot. Tackles https://github.com/ahrefs/monorobot/issues/154 and https://github.com/ahrefs/monorobot/issues/155.

When a PR is
- created -> posts message to the channel with the PR info + labels info (this becomes a thread for the PR)
- labeled -> will only show labels added post PR creation, will show them on the thread
- commented -> will add complete notification to the thread + summary notification on the channel with link to the thread
- reviewed -> same as above ☝️ + individual msg to the thread per comment in the review
- closed/merged -> posts message to the thread + summary notification to the channel + deletes state for that thread for cleanup

### CODE REVIEW

File: lib/api_local.ml [hunk 1]
```diff
@@ -88,6 +89,12 @@ module Slack : Api.Slack = struct
   let send_auth_test ~ctx:_ () =
     Lwt.return
     @@ Ok ({ url = ""; team = ""; user = ""; team_id = ""; user_id = "test_slack_user" } : Slack_t.auth_test_res)
+
+  let get_thread_permalink ~ctx:_ (thread : State_t.slack_thread) =
+    Lwt.return_some
+    @@ Printf.sprintf "https://monorobot.slack.com/archives/%s/p%s?thread_ts=%s&cid=%s" thread.cid
```
Comment (Khady) [inline lib/api_local.ml:95]: is it the correct url?

**Comment by Khady** [comment]
is it waiting on anything?

File: lib/api_local.ml (hunk 1 above)
Comment (thatportugueseguy) [reply lib/api_local.ml:95]: no, this is a dummy url. The structure is correct, but the domain is not. i didn't want to use the real one. This is only used for tests, I don't think it matters.

**Comment by thatportugueseguy** [comment]
This is how it looks:
Screenshot 2024-10-15 at 11 08 59

I've added these boxes and arrows to make it easier to connect the messages on the channel and the thread. The messages in boxes are independent messages, the messages with arrows are messages that are sent at the same time as the independent ones.
Screenshot 2024-10-15 at 11 00 43

I've just noticed that I have an issue with the merge notification styles, I'm going to fix those

File: lib/api_local.ml (hunk 1 above)
Comment (Khady) [reply lib/api_local.ml:95]: Can put as a comment in code please?

**Comment by thatportugueseguy** [comment]
@Khady can we merge this?

File: lib/slack.ml [hunk 2]
```diff
@@ -125,9 +191,10 @@ let generate_pr_review_comment_notification ~slack_match_func notification chann
   let file =
     match comment.path with
     | None -> None
-    | Some a -> Some (sprintf "New comment by %s in <%s|%s>" sender.login comment.html_url a)
+    | Some a -> Some (sprintf "Commented in file <%s|%s>" comment.html_url a)
```
Comment (yasunariw) [inline lib/slack.ml:194]: this change is unrelated to the PR

File: lib/slack.ml [hunk 3]
```diff
@@ -77,20 +100,42 @@ let generate_pull_request_notification ~slack_match_func notification channel =
...
-let generate_pr_review_notification ~slack_match_func notification channel =
+let generate_pr_review_notification ~slack_match_func ~(ctx : Context.t) ~get_thread_permalink notification channel =
   let { action; sender; pull_request; review; repository } = notification in
   let ({ number; title; html_url; _ } : pull_request) = pull_request in
   let action_str =
     match action with
     | Submitted ->
       (match review.state with
-      | "commented" -> "commented on"
+      | "commented" -> "reviewed"
```
Comment (yasunariw) [inline lib/slack.ml:138]: the change is unrelated to the PR

File: lib/slack.ml [hunk 4]
```diff
@@ -63,12 +73,25 @@ let format_attachments ~slack_match_func ~footer ~body =
...
   let ({ body; title; html_url; labels; merged; _ } : pull_request) = pull_request in
   let action, body =
     match action with
-    | Opened | Ready_for_review -> "opened", body
+    | Opened | Ready_for_review ->
+      let labels_banner = show_labels labels in
+      ( "opened",
+        body
+        |> Option.map (fun body' ->
+               Option.map_default (fun labels' -> sprintf "%s\n%s" body' labels') body' labels_banner) )
```
Comment (yasunariw) [inline lib/slack.ml:94]: the change is unrelated to the PR

File: lib/slack.atd [hunk 5]
```diff
@@ -144,6 +146,13 @@ type auth_test_res = {
   user_id: string;
 }
 
+type permalink_res = {
+  ok: bool;
```
Comment (yasunariw) [inline lib/slack.atd:150]: the field is unnecessary, it is handled by http_response below
